docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+  when they change, to reflect current values
+- The CSI plugin pod adds these config maps and secrets as volumes to its pod
+  specification, thus ensuring auto-refresh of content when these are updated
+  by the CO environment
+- VolumeCreate requests come against a targeted Ceph cluster-id and pool,
+  enabling the CSI plugin to locate the appropriate configuration information
+  for the cluster and its secrets to act on the request
+  - **NOTE:** The cluster-id and pool information is specified in the
+    StorageClass for example
+  - **NOTE:** In the **future**, if it is desired that the CSI plugin CS
+    choose which cluster and pool a request should be served from, this
+    information can be omitted from the StorageClass
+- The *volume_id* carries information regarding the *cluster_id* and pool,
+  enabling *volume_id* based RPC requests to operate using the provided
+  configuration information against the respective cluster
+- Usual CSI spec based maps such as, *volume_context* and *publish_context*
+  continue to carry information regarding required aspects about the
+  provisioned volume to dependent RPCs
+- The NS also has access to the config maps and secrets to satisfy required node
+  level RPC requests
+
+### Pros
+
+- Addresses removal of the current config map that maintains list of
+  provisioned volumes
+- Provides the ability for the CSI plugin to support all RPCs, as required
+  information is present with the plugin across all Ceph clusters
+- Safeguards secrets and cluster configuration information access to only
+  required namespaces that need to manage and use it
+- Provides the ability to have a single Ceph-CSI plugin instance per CO
+  instance, even when the CO instance needs volumes provisioned from multiple
+  Ceph clusters
+- Also, does not take away the capability, where needed, to use multiple
+  Ceph-CSI plugin instances (specialized using provisioner name) per CO
+  instance
+- Leaves the flexibility for the StorageClass to still override secrets, if such
+  requirements arise
+  - **NOTE:** The credentials stored for a cluster and passed to the CSI
+    plugins CS and NS, should be able to list volumes/images created using
+    other secrets (IOW, should act like a supervisor/root) to correctly
+    support ListVolumes and ListSnapshots operations
+- Provides the ability to deal with multi-cluster deployments, and possible
+  best-fit provisioning in the **future**
+  - Enables the ability for the CSI plugin to act on Topology based requests
+  - Simplifies StorageClass definitions, leaving the choice of the
+    cluster/pool to the CSI plugin
+
+### Cons
+
+- Addition/Removal of Ceph cluster configuration, to the CSI plugin, requires
+  that the pod specification be updated, and thus the CS and NS pods require
+  to be restarted, in order to pick up the latest cluster details
+  - If done right, the NS restarts should not impact existing volumes that
```

Comment:
Volumes mounted with ceph-fuse will have to be remounted...but there's little we can do about it ATM

Reply:
There is no restart required now, as the config file can and is passed in as a config map volume to the pod, thus there are no restarts.